Public/Src/Cache/ContentStore/Grpc/ContentStore.proto

```diff
@@ -33,6 +33,8 @@ service ContentServer {
 
     rpc CopyFile(CopyFileRequest) returns (stream CopyFileResponse) {}
 
+    rpc PushFile(stream PushFileRequest) returns (stream PushFileResponse) {}

```

Comment:
1. Why is this a bi-directional stream? 2. Why don't we use the same streaming chunk type for pushing and pulling? Here is proto I used in prototype:
service Copier {
  rpc Read (ReadRequest) returns (stream Chunk) {}
  rpc Write (stream Chunk) returns (WriteReply) {}
}

Reply:
I used a bidirectional stream to keep the whole transaction in a single request. In order to transmit the response headers as soon as we write them, it is required that the response type is a stream. Otherwise, the headers will be transmitted with the final response, making it impossible for us to complete the initial handshake to determine if copies are needed.

---
In reply to: [342196589](https://github.com/microsoft/BuildXL/pull/1141#discussion_r342196589) [](ancestors = 342196589)